x-pack/plugins/alerting/server/task_runner/schedule_actions_for_alerts.ts

```diff
@@ -0,0 +1,147 @@
+import { ExecutionHandler } from './create_execution_handler';
+import { ScheduleActionsForAlertsParams } from './types';
+import { AlertInstanceState, AlertInstanceContext } from '../types';
+import { Alert } from '../alert';
+import { RuleRunMetricsStore } from '../lib/rule_run_metrics_store';
+
+export async function scheduleActionsForAlerts<
+  InstanceState extends AlertInstanceState,
+  InstanceContext extends AlertInstanceContext,
+  ActionGroupIds extends string,
+  RecoveryActionGroupId extends string
+>(
+  params: ScheduleActionsForAlertsParams<
+    InstanceState,
+    InstanceContext,
+    ActionGroupIds,
+    RecoveryActionGroupId
+  >
+): Promise<void> {
+  const {
+    logger,
+    activeAlerts,
+    recoveryActionGroup,
+    recoveredAlerts,
+    executionHandler,
+    mutedAlertIdsSet,
+    ruleLabel,
+    ruleRunMetricsStore,
+    throttle,
+    notifyWhen,
+  } = params;
+  // execute alerts with executable actions
+  for (const [alertId, alert] of Object.entries(activeAlerts)) {
+    const executeAction: boolean = await shouldExecuteAction(
+      alertId,
+      alert,
+      mutedAlertIdsSet,
+      ruleLabel,
+      logger,
+      throttle,
+      notifyWhen
+    );
+    if (executeAction && alert.hasScheduledActions()) {
+      const { actionGroup, subgroup: actionSubgroup, state } = alert.getScheduledActionOptions()!;
+      await executeAlert(
+        alertId,
+        alert,
+        executionHandler,
+        ruleRunMetricsStore,
+        actionGroup,
+        state,
+        actionSubgroup
+      );
+    }
+  }
+
+  // execute recovered alerts
+  for (const alertId of Object.keys(recoveredAlerts)) {
+    if (mutedAlertIdsSet.has(alertId)) {
+      logger.debug(
+        `skipping scheduling of actions for '${alertId}' in rule ${ruleLabel}: instance is muted`
+      );
+    } else {
+      const alert = recoveredAlerts[alertId];
+      await executeAlert(
+        alertId,
+        alert,
+        executionHandler,
+        ruleRunMetricsStore,
+        recoveryActionGroup.id,
+        {} as InstanceState
+      );
+      alert.scheduleActions(recoveryActionGroup.id);
+    }
+  }
+}
+
+async function executeAlert<
+  InstanceState extends AlertInstanceState,
+  InstanceContext extends AlertInstanceContext,
+  ActionGroupIds extends string,
+  RecoveryActionGroupId extends string
+>(
+  alertId: string,
+  alert: Alert<InstanceState, InstanceContext, ActionGroupIds | RecoveryActionGroupId>,
+  executionHandler: ExecutionHandler<ActionGroupIds | RecoveryActionGroupId>,
+  ruleRunMetricsStore: RuleRunMetricsStore,
+  actionGroup: ActionGroupIds | RecoveryActionGroupId,
+  state: InstanceState,
+  actionSubgroup?: string
+) {
+  alert.updateLastScheduledActions(actionGroup, actionSubgroup);
+  alert.unscheduleActions();
+  return executionHandler({
+    actionGroup,
+    actionSubgroup,
+    context: alert.getContext(),
+    state,
+    alertId,
+    ruleRunMetricsStore,
+  });
+}
+
+async function shouldExecuteAction<
```

Comment:
does this function need to be async?

Reply:
oh no it doesn't, thanks for catching!

Reply:
Resolved in https://github.com/elastic/kibana/pull/138805/commits/13dfba87fe436349eefc1d2a21502508f95fd022